vowpalwabbit/core/src/reductions/csoaa_ldf.cc

```diff
@@ -493,111 +466,164 @@ void output_example(
       output_string_stream << costs[i].class_index << ':' << costs[i].partial_prediction;
     }
     // outputStringStream << std::endl;
-    all.print_text_by_ref(all.raw_prediction.get(), output_string_stream.str(), ec.tag, all.logger);
+    all.print_text_by_ref(output, output_string_stream.str(), ec->tag, logger);
   }
-
-  VW::details::print_cs_update(all, ec.l.cs.is_test_label(), ec, ec_seq, false, predicted_class);
 }
 
-void output_rank_example(VW::workspace& all, VW::example& head_ec, bool& hit_loss, VW::multi_ex* ec_seq)
+size_t cs_count_features(const VW::multi_ex& ec_seq)
 {
-  const auto& costs = head_ec.l.cs.costs;
-
-  if (VW::example_is_newline(head_ec)) { return; }
-  if (VW::is_cs_example_header(head_ec)) { return; }
-
-  all.sd->total_features += head_ec.get_num_features();
-
-  float loss = 0.;
-  VW::v_array<VW::action_score>& preds = head_ec.pred.a_s;
+  size_t num_features = 0;
+  for (const auto* ec : ec_seq)
+  {
+    if (VW::is_cs_example_header(*ec))
```

Comment:
won't the features have been merged by feature merger at this point? isn't it enough to just count the number of features in each example?